source/common/network/base_listener_impl.h

```diff
@@ -16,14 +16,18 @@ namespace Network {
  */
 class BaseListenerImpl : public virtual Listener {
 public:
-  BaseListenerImpl(Event::DispatcherImpl& dispatcher, Socket& socket);
+  /**
+   * @param socket the listening socket for this listener. It might be shared
+   * with other listeners if all listeners use single listen socket.
+   */
+  BaseListenerImpl(Event::DispatcherImpl& dispatcher, SocketSharedPtr socket);
```

Comment:
nit: passing `SocketSharedPtr&&` would be a bit more clear. Same in similar cases. That would likely satisfy @l8huang's readability concern also.

Reply:
OK that's fine.